File: 01-api-pedidos/manual/java/src/main/java/com/treinamento/api/controller/PedidoController.java
package com.treinamento.api.controller;

import com.treinamento.api.dto.PaginatedResponse;
import com.treinamento.api.dto.PedidoRequest;
import com.treinamento.api.model.Cliente;
import com.treinamento.api.model.ItemPedido;
import com.treinamento.api.model.Pedido;
import com.treinamento.api.model.Produto;
import com.treinamento.api.repository.ClienteRepository;
import com.treinamento.api.repository.ItemPedidoRepository;
import com.treinamento.api.repository.PedidoRepository;
import com.treinamento.api.repository.ProdutoRepository;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Sort;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/api/pedidos")
public class PedidoController {

    private static final List<String> STATUS_VALIDOS = List.of("pendente", "processando", "enviado", "entregue", "cancelado");

    private final PedidoRepository pedidoRepository;
    private final ClienteRepository clienteRepository;
    private final ProdutoRepository produtoRepository;
    private final ItemPedidoRepository itemPedidoRepository;

    public PedidoController(PedidoRepository pedidoRepository,
                            ClienteRepository clienteRepository,
                            ProdutoRepository produtoRepository,
                            ItemPedidoRepository itemPedidoRepository) {
        this.pedidoRepository = pedidoRepository;
        this.clienteRepository = clienteRepository;
        this.produtoRepository = produtoRepository;
        this.itemPedidoRepository = itemPedidoRepository;
    }

    @GetMapping
    public ResponseEntity<PaginatedResponse<Map<String, Object>>> listar(
            @RequestParam(required = false) String status,
            @RequestParam(required = false) String data_inicio,
            @RequestParam(required = false) String data_fim,
            @RequestParam(defaultValue = "1") int page,
            @RequestParam(defaultValue = "10") int limit) {

        PageRequest pageRequest = PageRequest.of(page - 1, limit, Sort.by(Sort.Direction.DESC, "id"));
        Page<Pedido> resultado = pedidoRepository.findAll(pageRequest);

        List<Map<String, Object>> data = resultado.getContent().stream()
                .map(this::pedidoToMapWithClienteNome)
                .toList();

        PaginatedResponse<Map<String, Object>> response = new PaginatedResponse<>(
                data,
                page,
                limit,
                resultado.getTotalElements()
        );

        return ResponseEntity.ok(response);
    }

    @GetMapping("/{id}")
    public ResponseEntity<?> buscarPorId(@PathVariable Long id) {
        return pedidoRepository.findById(id)
                .<ResponseEntity<?>>map(pedido -> {
                    Map<String, Object> response = new HashMap<>();
                    response.put("id", pedido.getId());
                    response.put("cliente_id", pedido.getCliente().getId());
                    response.put("status", pedido.getStatus());
                    response.put("total", pedido.getTotal());
                    response.put("created_at", pedido.getCreatedAt());
                    response.put("cliente_nome", pedido.getCliente().getNome());
                    response.put("cliente_email", pedido.getCliente().getEmail());
                    response.put("cliente_telefone", pedido.getCliente().getTelefone());

                    List<ItemPedido> itens = itemPedidoRepository.findByPedidoId(id);
                    List<Map<String, Object>> itensData = itens.stream()
                            .map(item -> {
                                Map<String, Object> i = new HashMap<>();
                                i.put("id", item.getId());
                                i.put("pedido_id", pedido.getId());
                                i.put("produto_id", item.getProduto().getId());
                                i.put("quantidade", item.getQuantidade());
                                i.put("preco_unitario", item.getPrecoUnitario());
                                i.put("produto_nome", item.getProduto().getNome());
                                return i;
                            })
                            .toList();
                    response.put("itens", itensData);

                    return ResponseEntity.ok(response);
                })
                .orElse(ResponseEntity.status(404).body(Map.of("error", "Pedido não encontrado")));
    }

    @PostMapping
    public ResponseEntity<?> criar(@RequestBody PedidoRequest request) {
        if (request.clienteId() == null) {
            return ResponseEntity.badRequest().body(Map.of("error", "Cliente não encontrado"));
        }

        Cliente cliente = clienteRepository.findById(request.clienteId()).orElse(null);
        if (cliente == null) {
            return ResponseEntity.badRequest().body(Map.of("error", "Cliente não encontrado"));
        }

        if (request.itens() == null || request.itens().isEmpty()) {
            return ResponseEntity.badRequest().body(Map.of("error", "Pedido deve ter pelo menos um item"));
        }

        // Validate all products exist and have stock
        List<Produto> produtosValidados = new ArrayList<>();
        for (PedidoRequest.ItemRequest itemReq : request.itens()) {
            Produto produto = produtoRepository.findById(itemReq.produtoId()).orElse(null);
            if (produto == null) {
                return ResponseEntity.badRequest().body(Map.of("error", "Um ou mais produtos não encontrados"));
            }
            if (produto.getEstoque() < (itemReq.quantidade() != null ? itemReq.quantidade() : 1)) {
                return ResponseEntity.badRequest().body(Map.of("error", "Estoque insuficiente para o produto " + produto.getNome()));
            }
            produtosValidados.add(produto);
        }

        double total = 0.0;
        for (int idx = 0; idx < request.itens().size(); idx++) {
            total += produtosValidados.get(idx).getPreco();
        }

        Pedido pedido = new Pedido();
        pedido.setCliente(cliente);
        pedido.setStatus("pendente");
        pedido.setTotal(total);
        Pedido salvo = pedidoRepository.save(pedido);

        // Insert items and decrement stock
        for (int idx = 0; idx < request.itens().size(); idx++) {
            PedidoRequest.ItemRequest itemReq = request.itens().get(idx);
            Produto produto = produtosValidados.get(idx);
            int quantidade = itemReq.quantidade() != null ? itemReq.quantidade() : 1;

            ItemPedido item = new ItemPedido();
            item.setPedido(salvo);
            item.setProduto(produto);
            item.setQuantidade(quantidade);
            item.setPrecoUnitario(produto.getPreco());
            itemPedidoRepository.save(item);

            produto.setEstoque(produto.getEstoque() - quantidade);
            produtoRepository.save(produto);
        }

        // Return bare pedido (no items, no client info) — matches Bun behavior
        Map<String, Object> response = new HashMap<>();
        response.put("id", salvo.getId());
        response.put("cliente_id", salvo.getCliente().getId());
        response.put("status", salvo.getStatus());
        response.put("total", salvo.getTotal());
        response.put("created_at", salvo.getCreatedAt());

        return ResponseEntity.status(201).body(response);
    }

    @PatchMapping("/{id}/status")
    public ResponseEntity<?> atualizarStatus(
            @PathVariable Long id,
            @RequestBody Map<String, String> body) {

        String novoStatus = body.get("status");
        if (novoStatus == null || novoStatus.isBlank()) {
            return ResponseEntity.badRequest().body(Map.of("error", "Status é obrigatório"));
        }

        if (!STATUS_VALIDOS.contains(novoStatus)) {
            return ResponseEntity.badRequest().body(Map.of("error",
                    "Status inválido. Valores permitidos: " + String.join(", ", STATUS_VALIDOS)));
        }

        return pedidoRepository.findById(id)
                .<ResponseEntity<?>>map(pedido -> {
                    pedido.setStatus(novoStatus);
                    Pedido salvo = pedidoRepository.save(pedido);
                    return ResponseEntity.ok(pedidoToMapWithClienteNome(salvo));
                })
                .orElse(ResponseEntity.status(404).body(Map.of("error", "Pedido não encontrado")));
    }

    private Map<String, Object> pedidoToMapWithClienteNome(Pedido pedido) {
        Map<String, Object> map = new HashMap<>();
        map.put("id", pedido.getId());
        map.put("cliente_id", pedido.getCliente().getId());
        map.put("status", pedido.getStatus());
        map.put("total", pedido.getTotal());
        map.put("created_at", pedido.getCreatedAt());
        map.put("cliente_nome", pedido.getCliente().getNome());
        return map;
    }
}

File: 01-api-pedidos/manual/php/src/Controllers/PedidoController.php
<?php

declare(strict_types=1);

namespace App\Controllers;

use App\Database;
use PDO;
use Psr\Http\Message\ResponseInterface as Response;
use Psr\Http\Message\ServerRequestInterface as Request;

class PedidoController
{
    private PDO $db;

    private const STATUS_VALIDOS = ['pendente', 'processando', 'enviado', 'entregue', 'cancelado'];

    public function __construct(?PDO $db = null)
    {
        $this->db = $db ?? Database::getInstance();
    }

    public function index(Request $request, Response $response): Response
    {
        $params = $request->getQueryParams();
        $page = max(1, (int)($params['page'] ?? 1));
        $limit = max(1, min(100, (int)($params['limit'] ?? 10)));
        $offset = ($page - 1) * $limit;


        $countStmt = $this->db->query("SELECT COUNT(*) FROM pedidos");
        $total = (int)$countStmt->fetchColumn();
        $totalPages = intdiv($total, $limit);

        $stmt = $this->db->prepare(
            "SELECT p.*, c.nome AS cliente_nome
             FROM pedidos p
             JOIN clientes c ON c.id = p.cliente_id
             ORDER BY p.id DESC
             LIMIT :limit OFFSET :offset"
        );
        $stmt->bindValue(':limit', $limit, PDO::PARAM_INT);
        $stmt->bindValue(':offset', $offset, PDO::PARAM_INT);
        $stmt->execute();
        $pedidos = $stmt->fetchAll();

        $payload = json_encode([
            'data'       => $pedidos,
            'page'       => $page,
            'limit'      => $limit,
            'total'      => $total,
            'totalPages' => $totalPages,
        ]);

        $response->getBody()->write($payload);
        return $response->withHeader('Content-Type', 'application/json');
    }

    public function show(Request $request, Response $response, array $args): Response
    {
        $id = (int)$args['id'];

        $stmtPedido = $this->db->prepare(
            "SELECT p.*, c.nome AS cliente_nome, c.email AS cliente_email, c.telefone AS cliente_telefone
             FROM pedidos p
             JOIN clientes c ON c.id = p.cliente_id
             WHERE p.id = :id"
        );
        $stmtPedido->execute([':id' => $id]);
        $pedido = $stmtPedido->fetch();

        if (!$pedido) {
            $response->getBody()->write(json_encode(['error' => 'Pedido não encontrado']));
            return $response->withHeader('Content-Type', 'application/json')->withStatus(404);
        }

        $stmtItens = $this->db->prepare(
            "SELECT ip.*, pr.nome AS produto_nome
             FROM itens_pedido ip
             JOIN produtos pr ON pr.id = ip.produto_id
             WHERE ip.pedido_id = :pedido_id"
        );
        $stmtItens->execute([':pedido_id' => $id]);
        $pedido['itens'] = $stmtItens->fetchAll();

        $response->getBody()->write(json_encode($pedido));
        return $response->withHeader('Content-Type', 'application/json');
    }

    public function create(Request $request, Response $response): Response
    {
        $body = $request->getParsedBody();

        $clienteId = $body['cliente_id'] ?? null;
        $itens = $body['itens'] ?? null;

        // Validate client_id exists
        if ($clienteId === null) {
            $response->getBody()->write(json_encode(['error' => 'Cliente não encontrado']));
            return $response->withHeader('Content-Type', 'application/json')->withStatus(400);
        }

        $clienteId = (int)$clienteId;

        $stmtCliente = $this->db->prepare("SELECT id FROM clientes WHERE id = :id");
        $stmtCliente->execute([':id' => $clienteId]);
        if (!$stmtCliente->fetch()) {
            $response->getBody()->write(json_encode(['error' => 'Cliente não encontrado']));
            return $response->withHeader('Content-Type', 'application/json')->withStatus(400);
        }

        if (!is_array($itens) || count($itens) === 0) {
            $response->getBody()->write(json_encode(['error' => 'Pedido deve ter pelo menos um item']));
            return $response->withHeader('Content-Type', 'application/json')->withStatus(400);
        }

        // Validate all products exist and have stock
        $produtoIds = array_map(fn($item) => (int)$item['produto_id'], $itens);
        $placeholders = implode(', ', array_fill(0, count($produtoIds), '?'));
        $stmtProds = $this->db->prepare(
            "SELECT id, preco, estoque, nome FROM produtos WHERE id IN ({$placeholders})"
        );
        $stmtProds->execute($produtoIds);
        $produtos = $stmtProds->fetchAll();

        if (count($produtos) !== count($produtoIds)) {
            $response->getBody()->write(json_encode(['error' => 'Um ou mais produtos não encontrados']));
            return $response->withHeader('Content-Type', 'application/json')->withStatus(400);
        }

        $produtoMap = [];
        foreach ($produtos as $p) {
            $produtoMap[(int)$p['id']] = $p;
        }

        // Check stock for each item
        foreach ($itens as $item) {
            $produtoId = (int)$item['produto_id'];
            $quantidade = (int)$item['quantidade'];
            $produto = $produtoMap[$produtoId] ?? null;

            if (!$produto) {
                $response->getBody()->write(json_encode(['error' => "Produto {$produtoId} não encontrado"]));
                return $response->withHeader('Content-Type', 'application/json')->withStatus(400);
            }

            if ((int)$produto['estoque'] < $quantidade) {
                $response->getBody()->write(json_encode(['error' => "Estoque insuficiente para o produto {$produto['nome']}"]));
                return $response->withHeader('Content-Type', 'application/json')->withStatus(400);
            }
        }

        $total = 0.0;
        foreach ($itens as $item) {
            $produto = $produtoMap[(int)$item['produto_id']];
            $total += (float)$produto['preco'];
        }

        // Insert pedido
        $stmtPedido = $this->db->prepare(
            "INSERT INTO pedidos (cliente_id, status, total) VALUES (:cliente_id, 'pendente', :total)"
        );
        $stmtPedido->execute([':cliente_id' => $clienteId, ':total' => $total]);
        $pedidoId = (int)$this->db->lastInsertId();

        // Insert items and decrement stock
        $stmtItem = $this->db->prepare(
            "INSERT INTO itens_pedido (pedido_id, produto_id, quantidade, preco_unitario) VALUES (:pedido_id, :produto_id, :quantidade, :preco_unitario)"
        );
        $stmtUpdateStock = $this->db->prepare(
            "UPDATE produtos SET estoque = estoque - :quantidade WHERE id = :id"
        );

        foreach ($itens as $item) {
            $produtoId = (int)$item['produto_id'];
            $quantidade = (int)$item['quantidade'];
            $produto = $produtoMap[$produtoId];

            $stmtItem->execute([
                ':pedido_id'     => $pedidoId,
                ':produto_id'    => $produtoId,
                ':quantidade'    => $quantidade,
                ':preco_unitario' => (float)$produto['preco'],
            ]);

            $stmtUpdateStock->execute([
                ':quantidade' => $quantidade,
                ':id'         => $produtoId,
            ]);
        }

        // Return bare pedido only (no items, no client info)
        $stmtGet = $this->db->prepare("SELECT * FROM pedidos WHERE id = :id");
        $stmtGet->execute([':id' => $pedidoId]);
        $novoPedido = $stmtGet->fetch();

        $response->getBody()->write(json_encode($novoPedido));
        return $response->withHeader('Content-Type', 'application/json')->withStatus(201);
    }

    public function updateStatus(Request $request, Response $response, array $args): Response
    {
        $id = (int)$args['id'];

        $stmtCheck = $this->db->prepare("SELECT id, status FROM pedidos WHERE id = :id");
        $stmtCheck->execute([':id' => $id]);
        $pedido = $stmtCheck->fetch();

        if (!$pedido) {
            $response->getBody()->write(json_encode(['error' => 'Pedido não encontrado']));
            return $response->withHeader('Content-Type', 'application/json')->withStatus(404);
        }

        $body = $request->getParsedBody();

        if (!isset($body['status'])) {
            $response->getBody()->write(json_encode(['error' => 'Campo obrigatório: status']));
            return $response->withHeader('Content-Type', 'application/json')->withStatus(400);
        }

        $novoStatus = (string)$body['status'];

        if (!in_array($novoStatus, self::STATUS_VALIDOS, true)) {
            $response->getBody()->write(json_encode([
                'error' => 'Status inválido. Valores permitidos: ' . implode(', ', self::STATUS_VALIDOS),
            ]));
            return $response->withHeader('Content-Type', 'application/json')->withStatus(400);
        }

        $stmtUpdate = $this->db->prepare("UPDATE pedidos SET status = :status WHERE id = :id");
        $stmtUpdate->execute([':status' => $novoStatus, ':id' => $id]);

        $stmtGet = $this->db->prepare("SELECT * FROM pedidos WHERE id = :id");
        $stmtGet->execute([':id' => $id]);
        $updated = $stmtGet->fetch();

        $response->getBody()->write(json_encode($updated));
        return $response->withHeader('Content-Type', 'application/json');
    }
}
